File: RAM.cpp
//
// Created by alexs on 3/15/2025.
//

#include "RAM.h"

#include <random>
#include <vector>
#include <ctime>
#include <strings.h>

RAM::RAM() {
    for (bool &i : free) {
        i = true;
    }
}

int RAM::pickFreeBlockIndex() {
    std::vector<int> freeIndices;

    for (int i = 0; i < RM_RAM_SIZE; ++i) {
        if (free[i]) {
            freeIndices.push_back(i);
        }
    }

    if (freeIndices.empty()) {
        return -1;
    }

    static std::mt19937 rng(static_cast<unsigned int>(std::time(nullptr)));
    std::uniform_int_distribution<size_t> dist(0, freeIndices.size() - 1);

    return freeIndices[dist(rng)];
}

int RAM::pickRandomBlockIndex() {
    random_device rd;
    mt19937 rng(rd());
    std::uniform_int_distribution<size_t> dist(0, RM_RAM_SIZE - 1);

    return dist(rng);
}

void RAM::writeWord(Word word, int block, int index) {
    free[block] = false;
    blocks[block].data[index] = word;
}

Word& RAM::readWord(int block, int index) {
    return blocks[block].data[index];
}


void RAM::printBlock(int block) {
    for (int i = 0; i < BLOCK_SIZE; i++) {
        printf("%.6s ", blocks[block].data[i].word);
    }
    printf("\n");
}

MemoryBlock *RAM::getBlock(int index) {
    return &blocks[index];
}

bool RAM::hasFreeSpace() {
    for (bool &i : free) {
        if (i) {
            return true;
        }
    }
    return false;
}

File: src/Machine/RAM.java
package Machine;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class RAM {
    private final boolean[] free = new boolean[Constants.RM_RAM_SIZE];
    private final MemoryBlock[] blocks = new MemoryBlock[Constants.RM_RAM_SIZE];
    private final Random rng = new Random(System.currentTimeMillis());

    public RAM() {
        for (int i = 0; i < Constants.RM_RAM_SIZE; i++) {
            free[i] = true;
            blocks[i] = new MemoryBlock();
        }
    }

    public void writeWord(Word word, int block, int index) {
        free[block] = false;
        blocks[block].data[index] = word;
    }

    public Word readWord(int block, int index) {
        return blocks[block].data[index];
    }

    public void printBlock(int block) {
        for (int i = 0; i < Constants.BLOCK_SIZE; i++) {
            System.out.printf("%.6s ", blocks[block].data[i].toString());
        }
        System.out.println();
    }

    public MemoryBlock getBlock(int index) {
        return blocks[index];
    }

    public int pickFreeBlockIndex() {
        List<Integer> freeIndices = new ArrayList<>();

        for (int i = 0; i < Constants.RM_RAM_SIZE; i++) {
            if (free[i]) {
                freeIndices.add(i);
            }
        }

        if (freeIndices.isEmpty()) {
            return -1;
        }

        return freeIndices.get(rng.nextInt(freeIndices.size()));
    }
}
